mypy/join.py

```diff
@@ -55,18 +55,14 @@ def join_instances(self, t: Instance, s: Instance) -> ProperType:
                     if not is_subtype(new_type, type_var.upper_bound):
                         self.seen_instances.pop()
                         return object_from_instance(t)
-                elif type_var.variance == CONTRAVARIANT:
-                    new_type = meet.meet_types(ta, sa)
-                    if len(type_var.values) != 0 and new_type not in type_var.values:
-                        self.seen_instances.pop()
-                        return object_from_instance(t)
-                    # No need to check subtype, as ta and sa already have to be subtypes of
-                    # upper_bound
-                elif type_var.variance == INVARIANT:
-                    new_type = join_types(ta, sa)
+                # TODO: contravariant case should use meet but pass seen instances as
+                # an argument to keep track of recursive checks.
+                elif type_var.variance in (INVARIANT, CONTRAVARIANT):
                     if not is_equivalent(ta, sa):
                         self.seen_instances.pop()
                         return object_from_instance(t)
+                    # Randomly choose the left one since they are equivalent.
+                    new_type = ta
```

Comment:
They could still be different, e.g. `list[str]` vs `list[Any]`? In this case the result should be `list[Any]`. I think that we'd need a recursive call here.